### PR TITLE
feat: enable Terraform debug mode via deployment configuration

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/8248

This PR mitigates the security concern around the untrusted Terraform plugin:
- It disables the debug mode by default
- It restricts access to the debug mode to admins only
- Debug mode can be enabled using deployment config:

```
CODER_ENABLE_TERRAFORM_DEBUG_MODE=true
```

Screenshot 2023-06-29 at 09 47 47

### CODE REVIEW

**Comment by johnstcn** [comment]
Should we:

a) Only allow admins to run builds in debug mode? (That is, hide the "try in debug mode" button unless you're an admin - this would resolve the issue with not being able to get the deployment values.)
b) Invert the debug mode logic - disable by default, and require users to enable? This is a more "secure by default" configuration.

**Review by johnstcn — COMMENTED** [review]
Change looks good to me mechanically, I just wonder if it's better to disable debug mode by default or only allow admins to use debug mode.

**Comment by mafredri** [comment]
I think @johnstcn makes some good points. I would take it even further such that enable debug mode is per-provisioner (disabled by default).

Another thought that comes to mind.. should the user even be allowed to see the build log from a debug-enabled build? I can foresee there being multiple knobs for this behavior. Like allow debug: never|admin|user. If set to admin, users shouldn't see the build log, etc.

**Comment by mtojek** [comment]
> I think @johnstcn makes some good points. 

I'm working on this 👍 

> Another thought that comes to mind.. should the user even be allowed to see the build log from a debug-enabled build? I can foresee there being multiple knobs for this behavior. Like allow debug: never|admin|user. If set to admin, users shouldn't see the build log, etc.

I would not like to increase the complexity of this PR as we have an issue to mitigate (disable unsafe logs). We can chat later about gaps we should address in the future. Here I would focus on enabling a global switch.

**Comment by mafredri** [comment]
> I would not like to increase the complexity of this PR as we have an issue to mitigate (disable unsafe logs). We can chat later about gaps we should address in the future. Here I would focus on enabling a global switch.

Agreed, that's fair. 👍🏻

**Comment by mtojek** [comment]
I refactored the PR considering the following statements:
- It disables the debug mode by default.
- It restricts access to the debug mode to admins only.
- Debug mode can be enabled using deployment config.

File: coderd/wsbuilder/wsbuilder.go [hunk 1]
```diff
@@ -638,11 +646,19 @@ func (b *Builder) authorize(authFunc func(action rbac.Action, object rbac.Object
 		}
 	}
 
-	if b.logLevel != "" && !authFunc(rbac.ActionUpdate, template) {
+	if b.logLevel != "" && !authFunc(rbac.ActionRead, rbac.ResourceDeploymentValues) {
```
Comment (mafredri) [inline coderd/wsbuilder/wsbuilder.go:649]: Is this the de-facto way to check for admin? It feels like it'd be one of those things that might change in the future (esp. a read permission). Probably fine for now though.

File: codersdk/deployment.go [hunk 2]
```diff
@@ -1215,10 +1216,20 @@ when required by your organization's security policy.`,
 			YAML:        "stackdriverPath",
 			Annotations: clibase.Annotations{}.Mark(annotationExternalProxies, "true"),
 		},
+		{
+			Name:        "Enable Terraform debug mode",
+			Description: "Enable debug mode while processing Terraform templates.",
```
Comment (mafredri) [inline codersdk/deployment.go:1221]:
```suggestion
			Description: "Allow administrators to enable debug mode while processing Terraform templates.",
```

> while processing Terraform templates.

Does it only apply to templates? Or is it workspace builds as well? Maybe something more generic like "Allow administrators to enable Terraform debug output."?

**Review by mafredri — APPROVED** [review]
Looks good!

File: coderd/wsbuilder/wsbuilder.go (hunk 1 above)
Comment (mtojek) [reply coderd/wsbuilder/wsbuilder.go:649]: Yes, that's why I covered it on the workspace build level with tests for all crucial roles.

File: codersdk/deployment.go (hunk 2 above)
Comment (mtojek) [reply codersdk/deployment.go]: Gotcha! I changed to the more generic form 👍

File: coderd/workspacebuilds_test.go [hunk 3]
```diff
@@ -640,11 +640,50 @@ func TestWorkspaceBuildStatus(t *testing.T) {
...
 	t.Parallel()
 
+	t.Run("DebugModeDisabled", func(t *testing.T) {
+		t.Parallel()
+
+		// Create user
+		deploymentValues := coderdtest.DeploymentValues(t)
+		deploymentValues.EnableTerraformDebugMode = false
+
+		templateAuthorClient := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true, DeploymentValues: deploymentValues})
```
Comment (johnstcn) [inline coderd/workspacebuilds_test.go:650]: nit: this is named misleadingly - this client is for the Owner.

File: codersdk/deployment.go [hunk 4]
```diff
@@ -1215,10 +1216,20 @@ when required by your organization's security policy.`,
 			YAML:        "stackdriverPath",
 			Annotations: clibase.Annotations{}.Mark(annotationExternalProxies, "true"),
 		},
+		{
+			Name:        "Enable Terraform debug mode",
+			Description: "Allow administrators to enable Terraform debug output.",
```
Comment (johnstcn) [inline codersdk/deployment.go:1221]: Is this closer to the intended behaviour?

```suggestion
			Description: "Allow administrators to enable Terraform debug output for template authors.",
```

**Review by johnstcn — APPROVED** [review]
👍

File: codersdk/deployment.go (hunk 4 above)
Comment (mtojek) [reply codersdk/deployment.go:1221]: Actually, it is not. It is rather:

`Allow administrators to enable Terraform debug output for administrators.`

so I will stick to the current form.

File: coderd/workspacebuilds_test.go (hunk 3 above)
Comment (mtojek) [reply coderd/workspacebuilds_test.go]: Good catch, Cian!

**Pull request merged**